general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+      }
+      return tm;
+   }
+};
+
+}
+
+/// Template class for build KDTree with template parameters Tindex
+/// specifying the type utilized for indexing the points, Tfloat
+/// specifying a float type for representing the coordinates of the
+/// points, integer parameter ndim specifying the dimensionality of the
+/// space and template function Tnorm for evaluating the distance
+/// between two points.
+template <typename Tindex, typename Tfloat, size_t ndim=3, typename Tnorm=KDTreeNorms::Norm_l2<Tfloat,ndim> >
+class KDTree
+{
+public:
+
+   /// Structure defining a geometric point in the ndim-dimensional
+   /// space.
+   struct PointND
+   {
+
+      /// Geometric point constructor
+      PointND() {for (size_t i=0; i<ndim; i++) {xx[i]=Tfloat(0.0);}}
+
+      /// Coordinates of the point
+      Tfloat xx[ndim];
+   };
+
+   /// Structure defining a node in the KDTree.
+   struct NodeND
+   {
+
+      /// Defines a point in the ndim-dimensional space
+      PointND pt;
+
+      /// Defines the attached index
+      Tindex  ind;
+   };
+
+   /// Default constructor
+   KDTree() {}
+
+   /// Returns the spatial dimension of the points
+   int SpatialDimension() {return ndim;}
+
+   /// Data iterator
+   typedef typename std::vector<NodeND>::iterator iterator;
+
+   /// Returns iterator to beginning of the point cloud
+   iterator begin() {return data.begin();}
+
+   /// Returns iterator to the end of the point cloud
+   iterator end() {return data.end();}
+
+   /// Returns the size of the point cloud
+   size_t size() {return data.size();}
+
+   /// Clears the point cloud
+   void clear() { data.clear();}
+
+   /// Builds the KDTree. If the point cloud is modified the tree
+   /// needs to be rebuild by a new call to Sort().
+   void Sort() { SortInPlace(data.begin(),data.end(),0);}
+
+   /// Adds a new node to the point cloud
+   void AddPoint(PointND& pt, Tindex ii)
+   {
+      NodeND nd;
+      nd.pt=pt;
+      nd.ind=ii;
+      data.push_back(nd);
+   }
+
+   /// Adds a new node by 3 coordinates and an associated index
+   void AddPoint(Tfloat x, Tfloat y, Tfloat z, Tindex ii)
+   {
+      MFEM_ASSERT(ndim==3,"The spatial dimension for the KDTree should be 3!")
+      NodeND nd;
+      nd.pt.xx[0]=x;  nd.pt.xx[1]=y; nd.pt.xx[2]=z; nd.ind=ii;
+      data.push_back(nd);
+   }
+
+   /// Adds a new node by 2 coordinates and an associated index
+   void AddPoint(Tfloat x, Tfloat y, Tindex ii)
+   {
+      MFEM_ASSERT(ndim==2,"The spatial dimension for the KDTree should be 2!")
+      NodeND nd;
+      nd.pt.xx[0]=x;  nd.pt.xx[1]=y; nd.ind=ii;
+      data.push_back(nd);
+   }
+
+   /// Adds a new node by 1 coordinate and an associated index
+   void AddPoint(Tfloat x, Tindex ii)
+   {
+      MFEM_ASSERT(ndim==1,"The spatial dimension for the KDTree should be 1!")
+      NodeND nd;
+      nd.pt.xx[0]=x; nd.ind=ii;
+      data.push_back(nd);
+   }
+
+   /// Finds the nearest neighbour index
+   Tindex FindClosestPoint(PointND& pt)
```

Comment:
It works on the KDTree data structure, thus it is implemented as a member function. It can be implemented as function as well. I do not have any preferences here.